packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -104,14 +105,19 @@ export default async ({
     webhookBody: webhookBody || {},
     pluginName,
   })
-  await getDataStore().ready()
 
-  const state = store.getState()
-  const nodes = getNodes()
+  await getDataStore().ready()
 
-  warnForPluginsWithoutNodes(state, nodes)
+  // We only warn for plugins w/o nodes and delete stale nodes on the first sourcing.
+  if (initialSourcing) {
+    const state = store.getState()
+    const nodes = getNodes()
 
-  deleteStaleNodes(state, nodes)
+    warnForPluginsWithoutNodes(state, nodes)
 
-  store.dispatch(actions.apiFinished({ apiName: `sourceNodes` }))
+    deleteStaleNodes(state, nodes)
```

Comment:
I guess we want to add a TODO to implement this later on as I'm sure we want to do this until we remove away from touchNodes.

```suggestion
    deleteStaleNodes(state, nodes)
```

Reply:
implement what exactly?

Reply:
deleting stale nodes on every call of sourceNodes?

Reply:
I can't remember 😅 I also don't know what I was referring to. But shouldn't we keep dispatching the apifinished piece

Reply:
oh I see what you mean now — the `store.dispatch(actions.apiFinished({ apiName: `sourceNodes` }))` line — hmm yeah didn't mean to remove that. It didn't seem to break anything either haha but I'll add it back

Reply:
Great that you figured it out! 👍  We need that for Gatsby Cloud engines uploading

Reply:
Ah sweet 👍 — stab the approve button when you have a chance and lets get this in!